File: data structures and algorithms/cpp/08_050.cpp
#include <cstddef> // for nullptr

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    TreeNode* symmetrizeTree(TreeNode* root) {
        if (!root) {
            return nullptr;
        }
        
        // Transform left and right subtrees first
        root->left = symmetrizeTree(root->left);
        root->right = symmetrizeTree(root->right);
        
        // Check if subtrees are symmetric
        if (!isSymmetric(root->left, root->right)) {
            // Not symmetric, make them symmetric
            if (root->left && !root->right) {
                // Only left child exists, mirror it
                root->right = mirrorTree(root->left);
            } else if (root->right && !root->left) {
                // Only right child exists, mirror it
                root->left = mirrorTree(root->right);
            } else {
                // Both children exist but not symmetric
                // Choose left subtree as reference and mirror it
                root->right = mirrorTree(root->left);
            }
        }
        
        return root;
    }
    
private:
    bool isSymmetric(TreeNode* left, TreeNode* right) {
        if (!left && !right) {
            return true;
        }
        if (!left || !right) {
            return false;
        }
        if (left->val != right->val) {
            return false;
        }
        
        return isSymmetric(left->left, right->right) && 
               isSymmetric(left->right, right->left);
    }
    
    TreeNode* mirrorTree(TreeNode* node) {
        if (!node) {
            return nullptr;
        }
        
        TreeNode* mirrored = new TreeNode(node->val);
        mirrored->left = mirrorTree(node->right);
        mirrored->right = mirrorTree(node->left);
        
        return mirrored;
    }
};

File: data structures and algorithms/cs/08_050.cs
using System;

public class TreeNode
{
   public int Val { get; set; }
   public TreeNode Left { get; set; }
   public TreeNode Right { get; set; }

   public TreeNode(int x)
   {
      Val = x;
   }
}

public class Solution
{
   public TreeNode SymmetrizeTree(TreeNode root)
   {
      if (root == null)
      {
         return null;
      }

      root.Left = SymmetrizeTree(root.Left);
      root.Right = SymmetrizeTree(root.Right);

      if (!IsSymmetric(root.Left, root.Right))
      {
         if (root.Left != null && root.Right == null)
         {
            root.Right = MirrorTree(root.Left);
         }
         else if (root.Right != null && root.Left == null)
         {
            root.Left = MirrorTree(root.Right);
         }
         else
         {
            root.Right = MirrorTree(root.Left);
         }
      }

      return root;
   }

   private bool IsSymmetric(TreeNode left, TreeNode right)
   {
      if (left == null && right == null)
      {
         return true;
      }
      if (left == null || right == null)
      {
         return false;
      }
      if (left.Val != right.Val)
      {
         return false;
      }

      return IsSymmetric(left.Left, right.Right) &&
             IsSymmetric(left.Right, right.Left);
   }

   private TreeNode MirrorTree(TreeNode node)
   {
      if (node == null)
      {
         return null;
      }

      TreeNode mirrored = new TreeNode(node.Val);
      mirrored.Left = MirrorTree(node.Right);
      mirrored.Right = MirrorTree(node.Left);

      return mirrored;
   }
}